### PR TITLE
check that mem reservations are valid

### PR DESCRIPTION
**Pull request by andreeaflorescu**
There are two error conditions that we need to look for:
- a memory reservation entry is invalid
- there are overlapping memory reservations

To allow only valid memory reservations entries, the address and size
are no longer public, and the `FdtReserveEntry` can only be created via
a constructor that checks if the region is valid. This allows for
simplifications when checking that memory regions are not overlapping.

Fixes: https://github.com/rust-vmm/vm-fdt/issues/5

### CODE REVIEW

File: src/writer.rs [hunk 1]
```diff
@@ -84,11 +92,49 @@ pub struct FdtWriter {
...
+        self.address.cmp(&other.address)
+    }
+}
+
+// Returns true if there are any overlapping memory reservations.
+fn check_overlapping(mem_reservations: &[FdtReserveEntry]) -> Result<()> {
+    let mut mem_rsvmap_copy = mem_reservations.to_vec();
+    mem_rsvmap_copy.sort();
+    let overlapping = mem_rsvmap_copy.windows(2).any(|w| {
+        // The following add cannot overlap because we can only have
```
Comment (danielverkamp) [inline src/writer.rs:128]: Nit: "overlap" -> "overflow" in this comment

File: src/writer.rs [hunk 2]
```diff
@@ -84,11 +92,49 @@ pub struct FdtWriter {
...
+
+impl FdtReserveEntry {
+    /// Create a memory reservation for the FDT.
+    ///
+    /// # Arguments
+    ///
+    /// * address: Physical address of the beginning of the reserved region.
+    /// * size: Size of the reserved region in bytes.
+    pub fn new(address: u64, size: u64) -> Result<Self> {
+        if address.checked_add(size).is_none() {
```
Comment (danielverkamp) [inline src/writer.rs:109]: We should probably also consider `size == 0` to be invalid.

File: src/writer.rs [hunk 3]
```diff
@@ -804,14 +845,57 @@ mod tests {
...
+        // Check that regions that overlap return an error on new.
+        // Regions start exactly at the end location of the
+        let overlapping = [
+            FdtReserveEntry::new(0x3, 1).unwrap(), // this overlaps with
+            FdtReserveEntry::new(0x0, 1).unwrap(),
+            FdtReserveEntry::new(0x2, 2).unwrap(), // this one.
+        ];
+        let fdt = FdtWriter::new(&overlapping);
+        assert_eq!(fdt.unwrap_err(), Error::OverlappingMemoryReservations);
+    }
```
Comment (danielverkamp) [inline src/writer.rs:900]: It might also be good to have a test case for valid reservations that are directly contiguous without overlapping just to make sure no off-by-one errors sneak in later - e.g. {0, 1}, {1, 1}, {2, 1}.

File: src/writer.rs [hunk 4]
```diff
@@ -804,14 +845,57 @@ mod tests {
...
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x0, 10).unwrap()),
+            Ordering::Greater
+        );
+    }
+
+    #[test]
+    fn test_overlapping_mem_reservations() {
+        // Check that regions that overlap return an error on new.
+        // Regions start exactly at the end location of the
```
Comment (danielverkamp) [inline src/writer.rs:892]: This comment may be missing the end of the sentence.

**Review by danielverkamp — COMMENTED** [review]
Looks good overall to me, thanks! Just commented on a few minor details I noticed.

File: src/writer.rs (hunk 2 above)
Comment (andreeaflorescu) [reply src/writer.rs:109]: Good point, will add that check as well.

File: src/writer.rs (hunk 3 above)
Comment (andreeaflorescu) [reply src/writer.rs:900]: Sounds like a good idea, will add it.

File: src/writer.rs [hunk 5]
```diff
@@ -804,14 +845,84 @@ mod tests {
...
+        let fdt = FdtWriter::new(&overlapping);
+        assert_eq!(fdt.unwrap_err(), Error::OverlappingMemoryReservations);
+    }
+
+    #[test]
+    fn test_off_by_one_mem_rsv() {
+        // This test is for making sure we do not introduce off by one errors
+        // in the memory reservations checks.
+        let non_overlapping = [
+            FdtReserveEntry::new(0x0, 1).unwrap(), // this overlaps with
```
Comment (danielverkamp) [inline src/writer.rs:921]: Maybe a copy-pasted comment from the other test? This one doesn't actually overlap.

Comment (andreeaflorescu) [reply src/writer.rs:921]: Yes :( true

**Review by lauralt — APPROVED** [review]

**Comment by andreeaflorescu** [comment]
@danielverkamp could you take another look? I addressed all the remaining comments.

**Review by danielverkamp — APPROVED** [review]
